src/ir/auth_logic/souffle_emitter.h

```diff
@@ -0,0 +1,122 @@
+  // arguments).
+  // To prevent instances of negated and non-negated uses of the predicate
+  // from generating two declarations, the sign here is always positive.
+  Predicate PredToDeclaration(const Predicate& predicate) {
+    std::vector<std::string> decl_args;
+    for (int i = 0; i < predicate.args().size(); i++) {
+      decl_args.push_back(absl::StrCat("x", std::to_string(i)));
+    }
+    return Predicate(predicate.name(), decl_args, kPositive);
+  }
+
+  std::string EmitPredicate(const Predicate& predicate) {
+    // Whenever a new predicate is encountered, it is added to the set of
+    // declarations (which does not include duplicates because it is a set).
+    declarations_.insert(PredToDeclaration(predicate));
+    return absl::StrCat(predicate.sign() == kNegated ? "!" : "",
+                        predicate.name(), "(",
+                        absl::StrJoin(predicate.args(), ", "), ")");
+  }
+
+  std::string EmitAssertionInner(const Predicate& predicate) {
+    return absl::StrCat(EmitPredicate(predicate), ".");
+  }
+
+  std::string EmitAssertionInner(const DLIRCondAssertion& cond_assertion) {
+    std::string lhs = EmitPredicate(cond_assertion.lhs());
+    std::vector<std::string> rhs_translated;
+    for (auto arg : cond_assertion.rhs()) {
+      rhs_translated.push_back(EmitPredicate(arg));
+    }
+    return absl::StrCat(std::move(lhs), " :- ",
+                        absl::StrJoin(rhs_translated, ", "), ".");
+  }
+
+  std::string EmitAssertion(const DLIRAssertion& assertion) {
+    return std::visit([this](auto value) { return EmitAssertionInner(value); },
+                      assertion.GetValue());
+  }
+
+  std::string EmitProgramBody(const DLIRProgram& program) {
+    std::vector<std::string> body_translated;
+    for (auto assertion : program.assertions()) {
+      body_translated.push_back(EmitAssertion(assertion));
+    }
+    return absl::StrJoin(body_translated, "\n");
+  }
+
+  std::string EmitOutputs(const DLIRProgram& program) {
+    return absl::StrJoin(program.outputs(), "\n",
+                         [](std::string* out, const std::string& prog_out) {
+                           return absl::StrAppend(
+                               out, absl::StrCat(".output ", prog_out));
+                         });
+  }
+
+  std::string EmitDeclaration(const Predicate& predicate) {
+    std::string arguments = absl::StrJoin(
+        predicate.args(), ", ", [](std::string* out, const std::string& arg) {
+          return absl::StrAppend(out, absl::StrCat(arg, ": symbol"));
+        });
+    return absl::StrCat(".decl ", predicate.name(), "(", arguments, ")");
+  }
+
+  std::string EmitDeclarations() {
+    return absl::StrJoin(declarations_, "\n",
+                         [this](std::string* out, auto decl) {
+                           return absl::StrAppend(out, EmitDeclaration(decl));
+                         });
+  }
+
+  absl::btree_set<Predicate> declarations_;
```

Comment:
Why a `btree_set` instead of a `flat_hash_set`?

Reply:
For testing purposes, I need to make sure that the order of declarations is consistent and does not change from one run to the next. With a `flat_hash_set` (which is what it was before) a consistent order is not guaranteed (or at least it wasn't once I changed the hash to the abseil one instead of the the role-my-own hash). (I guess another solution is to keep it as a `flat_hash_set` and generate a vector from the set and then order that vector at the end, but that seems just worse than this).

Reply:
It's better on some dimensions, worse on others. Using a `btree_set`, as you have here, is very concise. Asymptotically, it is the same as using a `flat_hash_set` and doing a big sort at the end; `btree_set` does `O(log(N))` insertion for each of the `N` elements you insert and can naively print the contents in order, while a `flat_hash_map` would do `O(1)` insertion an then have to do an `O(N * log(N))` sort at the end.

The difference between the two is that `btree_set` has to do an online sort as the elements are inserted, while the `sort` at the end with the `flat_hash_set` is offline, after all elements are known. In general, offline algorithms do much better than online algorithms, even if their complexity looks the same.

The code for performing the sort would be pretty easy too; it would look something like this:

```C++
std::vector elements_vec(declarations_.size());
elements_vec.insert(
  elements_vec.end(), std::make_move_iterator(declarations.begin()), std::make_move_iterator(declarations.end()));
std::sort(elements_vec);
```

In the end, I think that the conciseness of your current solution and the fact that this isn't likely to be performance critical means that the `absl::btree_set` is good enough to stay. But it's not "just worse", it's actually a pretty close consideration. I'd say leave it and just add a comment above `declarations_` saying that online ordering is only needed for tests. That way, if we're looking for ways to speed this up in the future, we can more easily remember that we have this knob to turn.

Reply:
OK you are right. That's better. I did it that way.

Reply:
Looks good :+1: